looper/utils.py

```diff
@@ -0,0 +1,157 @@
+                _LOGGER.info("No section '%s' in file '%s', skipping",
+                             s, self.path)
+                continue
+            # Test each of the section's commands.
+            try:
+                # Is section's data a mapping?
+                commands_iter = section_data.items()
+                self._logger.debug("Processing section '%s' data "
+                                   "as mapping", s)
+                for name, command in commands_iter:
+                    failed = self._store_status(section=s, command=command,
+                                                name=name)
+                    self._logger.debug("Command '%s': %s", command,
+                                       "FAILURE" if failed else "SUCCESS")
+            except AttributeError:
+                self._logger.debug("Processing section '%s' data as list", s)
+                commands_iter = data[s]
+                for cmd_item in commands_iter:
+                    # Item is K-V pair?
+                    try:
+                        name, command = cmd_item
+                    except ValueError:
+                        # Treat item as command itself.
+                        name, command = "", cmd_item
+                    failed = self._store_status(section=s, command=command,
+                                                name=name)
+                    self._logger.debug("Command '%s': %s", command,
+                                       "FAILURE" if failed else "SUCCESS")
+
+
+    def _store_status(self, section, command, name):
+        """
+        Based on new command execution attempt, update instance's
+        data structures with information about the success/fail status.
+        Return the result of the execution test.
+        """
+        failed = fails(command, name)
+        # Store status regardless of its value in the instance's largest DS.
+        self.section_to_fail_by_command[section][command] = failed
+        if failed:
+            # Only update the failure-specific structures conditionally.
+            self.failures_by_section[section].append(command)
+            self.failures.add(command)
+        return failed
+
+
+    @property
+    def failed(self):
+        """
+        Determine whether *every* command succeeded for *every* config file
+        section that was validated during instance construction.
+
+        :return bool: conjunction of execution success test result values,
+            obtained by testing each executable in every validated section
+        """
+        # This will raise exception even if validation was attempted,
+        # but no sections were used. Effectively, delegate responsibility
+        # to the caller to initiate validation only if doing so is relevant.
+        if not self.section_to_fail_by_command:
+            raise ValueError("No commands validated")
+        return 0 == len(self.failures)
+
+
+
+def fails(command, name=""):
```

Comment:
well, maybe: it should be positive, not negative (*does command succeed?* not *does command fail?*) and then can be named, if standalone, as `is_command_callable`; if part of a `Command` class then it could be just `is_callable`